pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+			log::trace!("Role edited: {:?}", role_hash);
+
+			Ok(())
+		}
+
+		/// Delete a **Role**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn delete_role(
+			origin: OriginFor<T>,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			cluster_id: Hash128,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+
+			// only the owner of the cluster can do this operation
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			let role_name_vec = role_name.into_inner();
+
+			// Check that the role exists in the cluster and in storage
+			if !cluster
+				.roles
+				.iter()
+				.any(|role| role_name_vec.eq(&role.name))
+			{
+				return Err(Error::<T>::RoleNotFound.into())
+			}
+
+			// write
+			// Remove Role from Cluster
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster
+					.roles
+					.iter()
+					.position(|x| x.name == role_name_vec)
+					.map(|index| cluster.roles.remove(index));
+			});
+
+			let role_hash = blake2_128(&[&cluster_id[..], &role_name_vec.as_slice()].concat());
+			<RoleToSettings<T>>::remove(&role_hash);
+
+			Self::deposit_event(Event::RoleDeleted { role_hash });
+			log::trace!("Role deleted: {:?}", role_hash);
+
+			Ok(())
+		}
+
+		/// Create a **Member**, give it a set of existing **Role** and assign it to an existing **Cluster**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn add_member(
+			origin: OriginFor<T>,
+			cluster_id: Hash128,
+			roles: Vec<Vec<u8>>,
+			member: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Check that the cluster exists and the caller is the owner of the cluster
+			let cluster = <Clusters<T>>::get(cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			// Check that the cluster does not already contain the member
+			ensure!(!cluster.members.contains(&member), Error::<T>::MemberExists);
+
+			ensure!(
+				cluster.members.len() < T::MembersLimit::get() as usize,
+				Error::<T>::MembersLimitReached
+			);
+
+			// Check that the roles for the member already exists in the cluster
+			let roles_in_cluster: Vec<Vec<u8>> =
+				cluster.roles.iter().map(|role| role.name.clone()).collect();
+
+			for role in &roles {
+				ensure!(roles_in_cluster.contains(&role), Error::<T>::RoleNotFound);
+			}
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				// Add member into the cluster
+				cluster.members.push(member.clone());
+
+				// Associate the member with its roles in the cluster
+				for role in roles {
+					cluster
+						.roles
```

Comment:
optimize

Reply:
This one might be harder to oprimize true, but BTMap might help

Reply:
I thought about BTMap, but discarded since the order of the elements is not important in this case and search is slower than Vec (O(log(n)) VS O(1)).
But i will look into it.

Reply:
@aang114 if `cluster.roles` becomes a `BTreeMap<role.name, Vec<AccountId>>` then the `Role` struct needs to be stored somewhere else. I cannot use `Role` as key like `BTreeMap<Role, Vec<AccountId>> `. So, this option implies the creation of a `StorageMap<cluster_Id, Role>`.

Reply:
What about
```rust
struct Role<AccountId> {
   pub members: Vec<AccountId>,
   pub rules: Option<Role>,
}
```

and so `cluster.roles` is: `BTreeMap<role.name, Role<AccountId>>`. 

Does that work?

Reply:
Same problem, `Role` is not stored anywhere. At this moment `Role` has not its own storage, it is included in `Clusters` storage itself (being part of `Cluster` struct as `Vec<Role>`).

Reply:
A solution could be to have 
```rust
cluster.roles: BTreeMap<role.name, Role>
cluster.members: BTreeMap<AccountId, role.name>
```
It would make the job, but it feels weird.

I already thought about all this at the beginning of this implementation for `pallet-clusters`, I created storage like `ClusterToRole<role_hash, Role>`, but then I remembered that for our chain the priority should be put on reducing the storage rather than having a super optimized computational complexity, and so I removed them and ended up with this code.

Reply:
Yup, I know 🙂. I'm suggesting that we should replace `Vec<Role>` to a `BTreeMap<role.name, Role<AccountId>>` ? (where `Role` is now defined in the comment above)

Reply:
Let me try with that thanks.

Reply:
It would be interesting to discuss a bit deeper about the tradeoff storage VS compexity. In pallet-clusters I do not think that the numbers will be very high (i.e. a user unlikely will have thousands of clusters, and so a single cluster is unlikely to have thousands of members). But instead, in pallet-protos and pallet-fragments, that storage mentioned before is very likely to skyrocket pretty quickly since they are just indexes.

Reply:
> Though, in pallet-fragments and pallet-protos we also have storage that has no functional meaning, but they allow to avoid iterations (such as fragments.DataHashMap, fragments.DataHashMapIndex or protos.MetaKeys, protos.MetaKeysIndex).
So, where is the boundary?

This is a good question, we have them for a few reasons:
1. To facilitate RPCs but not too much
2. To facilitate extrinsic functionality.

The issue here is that there will be a little amount of roles so simply it's not worth wasting time. As opposed to fragments.
I think this can stay as is even, all I want is best performance possible and correct code without any inefficient path.

Reply:
sure, I understand that. OK, I see your points.